openedx_events/content_authoring/data.py

```diff
@@ -0,0 +1,64 @@
+"""
+Data attributes for events within the architecture subdomain ``content_authoring``.
+
+These attributes follow the form of attr objects specified in OEP-49 data
+pattern.
+
+The attributes for the events come from the CourseDetailView in the LMS, with some unused fields removed
+(see deprecation proposal at https://github.com/openedx/public-engineering/issues/160)
+"""
+from datetime import datetime
+
+import attr
+from opaque_keys.edx.keys import CourseKey
+
+
+@attr.s(frozen=True)
+class CourseScheduleData:
+    """
+    Data describing course scheduling.
+
+    Arguments:
+        start (datetime): course start date
+        end (datetime): course end date
+        pacing (str): 'instructor' or 'self'
+        enrollment_start (datetime): start of course enrollment (optional)
+        enrollment_end (datetime): end of course enrollment (optional)
+    """
+
+    start = attr.ib(type=datetime)
+    end = attr.ib(type=datetime)
```

Comment:
Is `end` `None` when courses are self-paced?

Reply:
If so, you may want to make the type `Optional[datetime]` to indicate that `None` is a legal value.

Reply:
Sometimes. It's unfortunately rather variable.

Reply:
Start and end are documented as "optional", so should probably use `default=None`?

@ormsbee: See this ADR and the final section around punting on implementing `Optional`: https://github.com/openedx/openedx-events/blob/main/docs/decisions/0007-optional-fields.rst#deferredrejected-decisions Feel free to disagree, but that is where this was discussed.

Reply:
forgot about that. I'll make the end date optional. The start date should be kept mandatory since I'm fairly certain it's only nullable on legacy courses.

Reply:
I hadn't seen the ADR linked (thank you!), or maybe I saw it but forgot. I had personally been differentiating between "these are legal permutations of the data" (`type=Optional[datetime]`) vs. "we offer these convenient defaults for backwards compatibility or to make message creation easier". It looks like those are purposely conflated in ADR-0007. I have to think about it more, but my only longer term concern about that would be whether it complicates mypy style checking of type information. 

I'm also a bit confused about the wording of:

> We are also deferring any work around making an explicit `Optional` keyword for attrs fields

Is this referring to using `Optional` from the Python's `typing` module, or is it implying that there would have to create something new for some reason...?

Anyhow, these seem reversible decisions, and I definitely don't want to hold up this PR in order to re-hash an ADR. I withdraw my suggested change.

Reply:
Sounds like this conversation is resolved? Marking it as such.

Reply:
Re: ADR and `Optional`:
I'm guessing we were not clear other than to want to put this off. I'm guessing the ADR could more clearly state that `Optional` could be from Python's `typing` module, but it is still unclear how this all would work.